src/scenic/core/shapes.py

```diff
@@ -0,0 +1,151 @@
+
+    Represents a physical shape in Scenic. Does not encode position or orientation,
+    which are handled by the `Region` class. Does contain dimension information, which
+    is used as a default value by any `Object` with this shape and can be overwritten.
+
+    Args:
+        dimensions: The raw (before scaling) dimensions of the shape. If dimensions
+          and scale are both specified the dimensions are first set by dimensions, and then
+          scaled by scale.
+        scale: Scales all the dimensions of the shape by a multiplicative factor.
+          If dimensions and scale are both specified the dimensions are first set by dimensions,
+          and then scaled by scale.
+    """
+    def __init__(self, dimensions, scale):
+        # Store values
+        self.raw_dimensions = dimensions
+        self.scale = scale
+        self.dimensions = tuple(dim * self.scale for dim in self.raw_dimensions)
+        self.width = self.dimensions[0]
+        self.length = self.dimensions[1]
+        self.height = self.dimensions[2]
+
+    @cached_property
+    def containsCenter(self):
+        """Whether or not this object contains its central point"""
+        pq = trimesh.proximity.ProximityQuery(self.mesh)
+        region_distance = pq.signed_distance([(0,0,0)])[0]
+
+        return region_distance > 0
+
+    @property
+    @abstractmethod
+    def mesh(self):
+        pass
+
+    @property
+    @abstractmethod
+    def isConvex(self):
+        pass
+
+###################################################################################################
+# 3D Shape Classes
+###################################################################################################
+
+class MeshShape(Shape):
+    """ A Shape subclass defined by a Trimesh object.
+
+    Args:
+        mesh: A trimesh.Trimesh mesh object.
+        dimensions: The raw (before scaling) dimensions of the shape. If dimensions
+          and scale are both specified the dimensions are first set by dimensions, and then
+          scaled by scale.
+        scale: Scales all the dimensions of the shape by a multiplicative factor.
+          If dimensions and scale are both specified the dimensions are first set by dimensions,
+          and then scaled by scale.
+        initial_rotation: A 3-tuple containing the yaw, pitch, and roll respectively to apply when loading
+          the mesh. Note the initial_rotation must be fixed.
+    """
+    def __init__(self, mesh, dimensions=None, scale=1, initial_rotation=None):
+        # Ensure the mesh is watertight so volume is well defined
+        if not mesh.is_watertight:
+            raise ValueError("A MeshShape cannot be defined with a mesh that does not have a well defined volume.")
+
+        # Copy mesh and center vertices around origin
+        self._mesh = mesh.copy()
+        self._mesh.vertices -= self._mesh.bounding_box.center_mass
+        self._isConvex = self._mesh.is_convex
```

Comment:
You could move this into `isConvex` and make that a cached property. Or if we expect to need to know for all shapes whether or not they are convex, then it could just be another argument to the `Shape` initializer.